### DISCUSSION TITLE
Externalize macro-related targets

### DISCUSSION THREAD

**Discussion by abidon** [pointfreeco/swift-dependencies#222] [General]
### Description

The `DependenciesMacros` target requires the `swift-syntax` package, and the number of projects using this package is growing fast.

The main issue is that each library has different `swift-syntax` version constraints and it can become difficult to use libraries together if they aren't maintained at the same speed.

For anyone who doesn't use `DependenciesMacros`, getting blocked by it at packages resolution can be quite frustrating 😅

For this reason I believe anything related to macros should be opt-in instead of forced:
- it could find its place in a separate repository
- or gated behind some flags in the Package.swift — if that's possible (but I don't think it is? except with environment variables maybe? but it would be annoying again to setup when working with Xcode)

I would be glad to contribute to the effort.

> _The fewest dependencies, the better._

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [ ] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-dependencies/issues) or [discussion](https://github.com/pointfreeco/swift-dependencies/discussions).

### Expected behavior

Being able to resolve dependencies without being impacted by targets that are optional and unused.

### Actual behavior

Package resolution fails for `swift-syntax` because of conflicts with other libraries.

### Steps to reproduce

1. Create a new project
2. Add `swift-dependencies`
3. Add `swift-syntax` version 600 or later (or any library using swift-syntax with incompatible version constraints, i.e. `swift-testing` 0.8 or later)
4. Witness package dependencies resolution fail 😈

### Destination operating system

Any

### Swift Compiler version information

```shell
swift-driver version: 1.109.2 Apple Swift version 6.0 (swiftlang-6.0.0.3.300 clang-1600.0.20.10)
Target: arm64-apple-macosx15.0
```

**Comment by mbrandonw** [comment]
Hi @abidon, our dependence on SwiftSyntax is quite relaxed:

https://github.com/pointfreeco/swift-dependencies/blob/76b98dea40d2b13ace67216cbb100cef673e8267/Package%40swift-5.9.swift#L25

Which version of SwiftSyntax are you targeting? Is it some pre-release version for Swift 6?

It would be quite a breaking change to do this, and it would be a lot more annoying to maintain. It may be more preferable for us to maintain pre-releases branches ourselves that match SwiftSyntax's branches. But before making any decisions it would be good to get more information on your situation.

Since this isn't an actual bug with the library I am going to convert it to a discussion. Let's continue the conversation over there.

**Reply by mbrandonw** [reply]
Hi @abidon, luckily the 600.0.0 pre-release of SwiftSyntax doesn't have any breaking changes, so we are in the processing of loosening our requirement even further. That should unblock anyone depending on our libraries and swift-testing.

**Reply by abidon** [reply]
Thanks for releasing 1.3.1 in such a short time 🙏

I can confirm that it does work for `swift-dependencies` and `swift-testing`.

However, when I add another dependency that also rely on `swift-syntax` (with compatible version constraints), Swift Package Manager crashes.

Apart from having `DependenciesMacro` in a separate repository (and I do get your points against it), there is nothing more you can do on your side. I will report the bug to the Swift Package Manager team.

The conclusion is that I'll keep using my "lite" fork of `swift-dependencies` where I removed everything related to `DependeciesMacro` and `swift-syntax` for now until Swift Package Manager gets a bugfix.

**Reply by abidon** [reply]
Well, you guys are on all fronts 😅

Stephen already [reported the bug](https://github.com/apple/swift-package-manager/issues/7658).
